### PR TITLE
Prevent concurrent modification during set user properties

### PR DESCRIPTION
**Pull request by djih**

### CODE REVIEW

File: src/com/amplitude/api/AmplitudeClient.java [hunk 1]
```diff
@@ -564,11 +564,20 @@ public void setUserProperties(final JSONObject userProperties, final boolean rep
         runOnLogThread(new Runnable() {
             @Override
             public void run() {
-                Iterator<?> keys = userProperties.keys();
+                // Create deep copy of input userProperties to prevent ConcurrentModification
+                final JSONObject copy;
+                try {
+                    copy = new JSONObject(userProperties.toString());
```
Comment (curtisliu) [inline src/com/amplitude/api/AmplitudeClient.java]: Isn't there still a risk of ConcurrentModification cause you're calling toString here?

**Comment by djih** [comment]
Updated to make deep copy of input JSONObject, PTAL. Let me know if you want to catch or raise the ConcurrentModificationException

File: src/com/amplitude/api/AmplitudeClient.java [hunk 2]
```diff
@@ -548,12 +548,24 @@ public void setUserProperties(JSONObject userProperties) {
...
+        if (userProperties == null) {
+            if (replace) {
+                this.userProperties = null;
+            }
             return;
         }
 
-        if (userProperties == null) {
+        if (replace || this.userProperties == null) {
+            // Only create copy on main thread if necessary
```
Comment (curtisliu) [inline src/com/amplitude/api/AmplitudeClient.java]: Thinking about it the entire replacement should happen on the LogThread

Comment (djih) [inline src/com/amplitude/api/AmplitudeClient.java]: I was originally thinking about that too, but I was worried about the capture in the run. getInstance() grabs the wrong instance, but looks like I can capture "this" for the run.

**Comment by djih** [comment]
Moved all copying onto background thread, PTAL

**Comment by djih** [comment]
Moved all logic onto background thread, PTAL

**Comment by curtisliu** [comment]
LGTM

**Pull request merged**